### PR TITLE
Avoid cascading failure: give up on incoming HTLCs in time if outgoing is stuck.

### PR DESCRIPTION
**Pull request by rustyrussell**
@t-bast described this issue in general, and here's the fix, with some cleanups on the way.

1. We forward an HTLC.
2. The outgoing channel goes onchain
3. We try to spend the outgoing HTLC (onchain) to time it out.
4. But we either don't pay enough because fees spiked, or with anchors, we decide it's not worthwhile to pay enough fees.
5. HTLC stays open and incoming side gets upset, closing channel.

Our solution is to monitor for this: 

1. If an htlc is incoming, and <= 3 blocks away from timing out
2. And there's a corresponding outgoing
3. And it's onchain (it will be by default, since cltv_delta is 6 (testnet) or 34 (mainnet).
4. Fail the incoming immediately.

This is a risk, of course: the outgoing HTLC could be claimed by the peer.  But that's no worse to us than it not getting mined at all, which we were prepared for.

### CODE REVIEW

**Comment by rustyrussell** [comment]
Trivial rebase.

File: onchaind/onchaind.c [hunk 1]
```diff
@@ -1499,12 +1499,16 @@ static void handle_preimage(struct tracked_output **outs,
...
-			status_broken("HTLC already resolved by %s"
-				     " when we found preimage",
-				     tx_type_name(outs[i]->resolved->tx_type));
-			return;
+			if (outs[i]->resolved->tx_type != SELF) {
+				status_broken("HTLC already resolved by %s"
+					      " when we found preimage",
+					      tx_type_name(outs[i]->resolved->tx_type));
+				return;
+			}
```
Comment (vincenzopalazzo) [inline onchaind/onchaind.c:1511]: we can remove the `{..}` due the single stmt inside the if?

File: lightningd/peer_htlcs.c [hunk 2]
```diff
@@ -1754,8 +1754,14 @@ void onchain_failed_our_htlc(const struct channel *channel,
...
-		local_fail_in_htlc(hout->in,
-				   take(towire_permanent_channel_failure(NULL)));
+ 		/* Careful!  We might have already timed out incoming
+		 * HTLC in consider_failing_incoming */
+		if (hout->in->badonion == 0
+		    && !hout->in->failonion
+		    && !hout->in->preimage) {
+			local_fail_in_htlc(hout->in,
+					   take(towire_permanent_channel_failure(NULL)));
+		}
```
Comment (vincenzopalazzo) [inline lightningd/peer_htlcs.c:1764]: same here?

File: lightningd/peer_htlcs.c [hunk 3]
```diff
@@ -2661,6 +2667,43 @@ static u32 htlc_in_deadline(const struct lightningd *ld,
...
+		return;
+
+	/* Already done it once? */
+	if (hout->in->failonion)
+		return;
+
+	/* OK, if we're within 3 blocks of upstream getting upset, force it
+	 * to fail without waiting for onchaind. */
+	if (height + 3 < hout->in->cltv_expiry)
+		return;
```
Comment (vincenzopalazzo) [inline lightningd/peer_htlcs.c:2697]: Can we print some log lines at debug level? I feel that we are missing some helpful debug information

**Review by vincenzopalazzo — COMMENTED** [review]
LGTM, just some tiny comments on it but for me we are ready to go

File: lightningd/peer_htlcs.c (hunk 3 above)
Comment (rustyrussell) [reply lightningd/peer_htlcs.c:2697]: I did temporarily, but it's very noisy!  Every HTLC on every block, basically...

**Pull request merged**